### PR TITLE
WIP - Coredns user config

### PR DESCRIPTION
**Pull request by erikwilson**
For #462

Provides a `coredns-user` configMap to separate k3s config from user config.

Example:
```sh
cat >/var/lib/rancher/k3s/server/manifests/user-coredns-config.yaml <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns-user
  namespace: kube-system
data:
  Rewrite: |
    rewrite name regex [a-zA-Z.]+\.dev0\.be\.lan traefik.kube-system.svc.cluster.local
EOF
```

### CODE REVIEW

File: manifests/coredns.yaml [hunk 1]
```diff
@@ -55,23 +55,33 @@ data:
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: coredns-user
+  namespace: kube-system
+data:
+  EmptyExample: |
+    # This is an example empty coredns configmap entry to silence warnings
+---
 apiVersion: extensions/v1beta1
```
Comment (warmchang) [inline manifests/coredns.yaml]: apiVersion: apps/v1 is better for k8s-v1.14.x, :smiley:

**Comment by mrueg** [comment]
Can we decouple the image bump from the feature add? I would like to see a more recent version of coredns soon :)

**Comment by mlebrun** [comment]
@erikwilson appreciate you opening this! is there any way I can help in order to get this PR to move forward? Very excited and appreciate you taking time to do this!

**Comment by galal-hussein** [comment]
@erikwilson I am wondering if we should do this to all manifests, the local storage PR also include a configmap that can be configured for different nodes https://github.com/rancher/k3s/pull/816/files#diff-0683dbb4f157e849170880c2b3a8049eR94

**Comment by erikwilson** [comment]
Thanks @mlebrun! The PR kind of stalled due to limitations with CoreDNS config. I would like to support a `hosts` directive for users (or at least maintain our own node hosts independent of user hosts settings), but it appears that the hosts plugin can only be used once per zone. Perhaps the PR should move forward anyways tho if the basic usage looks okay.

Unfortunately @galal-hussein I think this is using some CoreDNS specific tricks using the `reload` and `import` directives so not sure how portable it is. It would be nice to find a better way to customize manifests tho, since naming and upgrades are an issue.

**Comment by mlebrun** [comment]
Right, if some version of this functionality could move forward, we would at least have something in the meantime, albeit maybe not perfect. But I would take that over having to modify my CoreDNS config on server restarts! As always, let me know if there's anything I can do outside of gentle nudging 😃

**Comment by rpocase** [comment]
@erikwilson Any plans to revisit this PR? This feature would be extremely valuable for my dev tooling.

**Comment by blaggacao** [comment]
Is this somehow superseded? (eg. by what's yet to come in https://github.com/rancher/k3s/pull/1899)

**Comment by ieugen** [comment]
Is this PR still relevant? 
https://github.com/rancher/k3s/pull/1899 has been merged.
If not maybe it can be closed.

**Comment by dereknola** [comment]
Closing this PR as out-of-date (by 2 years)
Native config file support exists. https://github.com/k3s-io/k3s/pull/1899

**Comment by ChristianCiach** [comment]
I dont't see how #1899 has any relation to this issue. I need to modify the coredns config to do this: https://coredns.io/2017/05/08/custom-dns-entries-for-kubernetes/

But the `config.yaml` doesn't help me with this, doesn't it?

**Comment by iwilltry42** [comment]
Coming here today as I'm struggling with finding a proper way of modifying the CoreDNS configmap for k3d (e.g. to inject host.k3d.internal and records for the LB and registries, etc.).
Up to now, we're modifying the `NodeHosts` field of the configmap (as the `hosts` plugin can only be used once per server), but this feels unsafe, as K3s is also updating this field every now and then.
It works ok so far, but I think having a separate field or even an extra configmap for user-defined hosts would be great for this :)

**Pull request closed without merging**